exercises/practice/dnd-character/dnd_character.vader

```diff
@@ -0,0 +1,95 @@
+
+Execute (ability modifier for score 3 is -4):
+  let g:score = 3
+  let g:expected = -4
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 4 is -3):
+  let g:score = 4
+  let g:expected = -3
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 5 is -3):
+  let g:score = 5
+  let g:expected = -3
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 6 is -2):
+  let g:score = 6
+  let g:expected = -2
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 7 is -2):
+  let g:score = 7
+  let g:expected = -2
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 8 is -1):
+  let g:score = 8
+  let g:expected = -1
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 9 is -1):
+  let g:score = 9
+  let g:expected = -1
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 10 is 0):
+  let g:score = 10
+  let g:expected = 0
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 11 is 0):
+  let g:score = 11
+  let g:expected = 0
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 12 is +1):
+  let g:score = 12
+  let g:expected = 1
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 13 is +1):
+  let g:score = 13
+  let g:expected = 1
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 14 is +2):
+  let g:score = 14
+  let g:expected = 2
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 15 is +2):
+  let g:score = 15
+  let g:expected = 2
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 16 is +3):
+  let g:score = 16
+  let g:expected = 3
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 17 is +3):
+  let g:score = 17
+  let g:expected = 3
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (ability modifier for score 18 is +4):
+  let g:score = 18
+  let g:expected = 4
+  AssertEqual g:expected, Modifier(g:score)
+
+Execute (random ability is within range):
+  let g:score = Ability()
+  Assert g:score >= 3 && g:score <= 18
+
+Execute (random character is valid):
+  let g:char = Character()
+  let g:invalid_stats = keys(filter(g:char, 'g:char[v:key] < 3 || g:char[v:key] > 19'))
+  AssertEqual 0, len(g:invalid_stats)
+
+Execute (each ability is only calculated once):
+  let g:char = Character()
+  let g:strength1 = g:char["strength"]
+  let g:strength2 = g:char["strength"]
+  AssertEqual g:strength1, g:strength2
```

Comment:
This does not check each ability, it only tests strength (and that is an attribute not an ability), the test name may be misleading.

Reply:
Looks like the Ability is coming from problem specifications, so if it is something that is confusing, it may need to be done up there (or we can fix it in here, and have it before Problem Specifications changes roll through.

Reply:
Yeah, there's an open PR about that in problem-specifications from two weeks ago. The test was added five years ago because a student was randomizing the ability score directly in the getter logic. It makes sense that you either do that for all the getters or you implement the actual dice roll logic. Why do both? So if you check one stat and they're not recalculated, presumably the rest are.

That said, yeah it makes more sense to just check all the abilities. It doesn't add that much overhead and that's the only way "each ability is only calculated once" can be proven.

Reply:
We probably could skip this test though. We're doing dictionary lookups directly instead of calling a getter function so I don't see how a student can intercept the outgoing value and replace it with a randomly-generated one.

Reply:
I decided to not include this test or its recent reimplementation.